### PR TITLE
[Feature | Refac] Add Tensor method & Symbolic HamiltonianEvolution

### LINKED ISSUE #179: [Refactoring, Hamevo] Port HamEvo logic from Qadence
Port the HamEvo logic from Qadence backend to native PyQ.

### PR DESCRIPTION
**Pull request by dominikandreasseitz**
Closes #179 

@dominikandreasseitz TODOS:
- [x] Create interface for `HamiltonianEvolution` to allow for symbolic (hamiltonians passed as parameters), tensor and parametric generators from qadence into pyq [based on the qadence implementation ](https://github.com/pasqal-io/qadence/blob/main/qadence/backends/pyqtorch/convert_ops.py#L291)
- [x] Create `tensor` method for `Primitive`, `Sequence` + subtypes which corresponds to [`block_to_tensor`](https://github.com/pasqal-io/qadence/blob/main/qadence/blocks/block_to_tensor.py)from qadence

@chMoussa TODOs:
- [ ] Add block_to_tensor tests based on https://github.com/pasqal-io/qadence/blob/main/tests/qadence/test_matrices.py
- [ ] Add matrix endianness tests based on https://github.com/pasqal-io/qadence/blob/main/tests/backends/test_endianness.py
- [ ] Make sure this branch passes all the qadence tests when run with https://github.com/pasqal-io/qadence/pull/406 

Optional:
- [ ] Support `use_full_qubit_support` option from qadence (ask @jpmoutinho)

### CODE REVIEW

File: pyqtorch/analog.py [hunk 1]
```diff
@@ -57,62 +79,132 @@ def __init__(self, operations: list[Module]):
...
 
-        def _diag_operator(hamiltonian: Operator, time_evolution: torch.Tensor) -> Operator:
-            evol_operator = torch.diagonal(hamiltonian) * (-1j * time_evolution).view((-1, 1))
-            evol_operator = torch.diag_embed(torch.exp(evol_operator))
-            return torch.transpose(evol_operator, 0, -1)
 
-        def _matrixexp_operator(hamiltonian: Operator, time_evolution: torch.Tensor) -> Operator:
-            evol_operator = torch.transpose(hamiltonian, 0, -1) * (-1j * time_evolution).view(
-                (-1, 1, 1)
+class HamiltonianEvolution(Sequence):
```
Comment (dominikandreasseitz) [inline pyqtorch/analog.py:130]: @Roland-djee @jpmoutinho my proposal for the new HamEvo which will handle everything we need in qadence. not finished yet. mainly block_to_tensor-related things have to be resolved still

**Comment by jpmoutinho** [comment]
Hey @chMoussa @dominikandreasseitz, regarding the porting of tests from https://github.com/pasqal-io/qadence/blob/main/tests/qadence/test_matrices.py, I think we should avoid porting the first ones on the projectors that run `block_to_tensor` on some block that ends up being more than a 3-qubit matrix and compares it to some pre-defined and manually written huge matrix. They just get very hard to navigate especially because of how the linter forces line breaks in the manually written matrices.

Testing this should just be the same workflow as done with the `_calc_mat_vec_wavefunction` function in the Qadence file: some random / specific block is composed, it is applied to some random initial state through the normal PyQ `apply_gate`, and then you matmul the tensor version of the block with the same initial state and check you get the same final state. This workflow should be complete enough to guarantee that the implementation is correct for any block composition.

File: pyqtorch/circuit.py [hunk 2]
```diff
@@ -1,21 +1,24 @@
...
 
 import logging
+from collections import Counter
 from functools import reduce
 from logging import getLogger
 from operator import add
 from typing import Any, Generator, Iterator, NoReturn
 
 import torch
+from numpy import int64
```
[the remark this answers is not in the document]
Comment (chMoussa) [reply pyqtorch/circuit.py:11]: Sometimes we see code/tests using numpy integers for the target, and numpy integers are not instances of python int. So I added a small if statement to take care of this and make tests passes.

File: pyqtorch/utils.py [hunk 3]
```diff
@@ -105,17 +105,20 @@ def is_normalized(state: Tensor, atol: float = ATOL) -> bool:
...
 
     Returns:
         True if diagonal, else False.
     """
-    return (
-        len(torch.abs(torch.triu(H, diagonal=1)).to_sparse().coalesce().values()) == 0
-    )
+    m = H.shape[0]
+    p, q = H.stride()
+    offdiag_view = torch.as_strided(H[:, 1:], (m - 1, m), (p + q, q))
```
Comment (dominikandreasseitz) [inline pyqtorch/utils.py:120]: @jpmoutinho

File: pyqtorch/utils.py [hunk 4]
```diff
@@ -77,7 +77,9 @@ def is_diag(H: Tensor) -> bool:
     """
     Returns True if Hamiltonian H is diagonal.
     """
-    return len(torch.abs(torch.triu(H, diagonal=1)).to_sparse().coalesce().values()) == 0
+    return (
+        len(torch.abs(torch.triu(H, diagonal=1)).to_sparse().coalesce().values()) == 0
+    )
```
[the remark this answers is not in the document]
Comment (dominikandreasseitz) [reply pyqtorch/utils.py]: replaced by a more efficient version by @chMoussa

File: pyqtorch/analog.py [hunk 5]
```diff
@@ -54,65 +82,177 @@ class Add(Sequence):
...
-class Hamiltonian(Add):
-    def __init__(self, operations: list[torch.nn.Module]):
-        if all([not isinstance(op, (Parametric)) for op in operations]):
-            super().__init__(operations)
-        else:
-            raise TypeError(
-                "Hamiltonian can only contain the following operations: [Primitive, Scale, Add]."
-            )
 
+class Observable(Sequence):
```
[the remark this answers is not in the document]
Comment (chMoussa) [reply pyqtorch/analog.py:149]: @jpmoutinho Thanks for pointing it out. We created the issue #192 for taking care of it in another MR as this one becomes tedious.

File: pyqtorch/utils.py [hunk 6]
```diff
@@ -75,9 +105,20 @@ def is_normalized(state: Tensor, atol: float = ATOL) -> bool:
...
+        H: Input tensor.
+
+    Returns:
+        True if diagonal, else False.
     """
-    return len(torch.abs(torch.triu(H, diagonal=1)).to_sparse().coalesce().values()) == 0
+    m = H.shape[0]
+    p, q = H.stride()
+    offdiag_view = torch.as_strided(H[:, 1:], (m - 1, m), (p + q, q))
+    return torch.count_nonzero(offdiag_view) == 0
```
Comment (jpmoutinho) [inline pyqtorch/utils.py:121]: Cool :)

Is this much faster than the previous version?

One thing that will have to be "fixed" is to account for floating point arithmetic errors, since you may end up with very small non-zero off-diagonal elements. I believe this is what caused the previous version to fail, and I think it will also fail in this case. Anyway, this can easily be done by replacing the `torch.count_nonzero` by some efficient check of greater than $\epsilon$.

File: pyqtorch/utils.py [hunk 7]
```diff
@@ -127,14 +217,27 @@ def _rand(n_qubits: int) -> Tensor:
         sumx = torch.sum(x)
         phases = torch.rand(N, generator=generator) * 2.0 * torch.pi
         return _normalize(
-            (torch.sqrt(x / sumx) * torch.exp(1j * phases)).reshape(N, 1).type(dtype).to(device)
+            (torch.sqrt(x / sumx) * torch.exp(1j * phases))
+            .reshape(N, 1)
+            .type(dtype)
+            .to(device)
```
Comment (jpmoutinho) [inline pyqtorch/utils.py:223]: Hmm possibly minor thing but shouldn't the `dtype` argument be passed to `torch.rand`? Otherwise it defaults to `float32`, and it could influence the precision of the final random state regardless of the precision set in the input.

I think in this case it's not really making a difference because the normalization seems to be correcting the final precision.

**Review by jpmoutinho — APPROVED** [review]
Thanks a lot guys 🙏

Definitely it will be easier to iron out the details in separate MRs.

File: pyqtorch/circuit.py [hunk 8]
```diff
@@ -62,22 +65,47 @@ def __init__(self, operations: list[Module]):
...
+            )
+        )
+
+        self._qubit_support = tuple(
+            map(
+                lambda x: x if isinstance(x, (int, int64)) else x[0],
+                self._qubit_support,
+            )
+        )
+        assert all([isinstance(q, (int, int64)) for q in self._qubit_support])
```
[the remark this answers is not in the document]
Comment (dominikandreasseitz) [reply pyqtorch/circuit.py:87]: there are various cases where we pass numpy ints from qadence. also we need numpy for apply_operator

File: pyqtorch/utils.py (hunk 6 above)
Comment (chMoussa) [reply pyqtorch/utils.py]: Yes, much faster when I tried. I just added a tolerance argument for such cases.

File: pyqtorch/circuit.py [hunk 9]
```diff
@@ -62,22 +65,47 @@ def __init__(self, operations: list[Module]):
...
+                sum(
+                    [
+                        op.qubit_support
+                        for op in self.operations
+                        if hasattr(op, "qubit_support")
+                    ],
+                    (),
+                )
+            )
+        )
```
[the remark this answers is not in the document]
Comment (dominikandreasseitz) [reply pyqtorch/circuit.py:79]: good catch, all should have that now

File: pyqtorch/analog.py [hunk 10]
```diff
@@ -43,34 +48,78 @@ def pre_backward_hook(*args, **kwargs) -> None:  # type: ignore[no-untyped-def]
...
+    ) -> Tensor:
+        thetas = (
+            values[self.param_name]
+            if isinstance(self.param_name, str)
+            else self.param_name
+        )
+        return thetas * self.operations[0].tensor(values, n_qubits, diagonal)
+
+    def flatten(self) -> list[Scale]:
+        return [self]  # we dont want to flatten this
```
[the remark this answers is not in the document]
Comment (dominikandreasseitz) [reply pyqtorch/analog.py]: added

File: pyqtorch/analog.py [hunk 11]
```diff
@@ -43,34 +48,78 @@ def pre_backward_hook(*args, **kwargs) -> None:  # type: ignore[no-untyped-def]
...
         return thetas * self.operations[0].unitary(values)
 
     def dagger(self, values: dict[str, Tensor]) -> Tensor:
         return _dagger(self.unitary(values))
 
     def jacobian(self, values: dict[str, Tensor]) -> Tensor:
-        return values[self.param_name] * ones_like(self.unitary(values))
+        raise NotImplementedError
+        # thetas = values[self.param] if isinstance(self.param, str) else self.param_name
+        # return thetas * ones_like(self.unitary(values))
```
[the remark this answers is not in the document]
Comment (dominikandreasseitz) [reply pyqtorch/analog.py:111]: i added context

File: pyqtorch/circuit.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (dominikandreasseitz) [reply pyqtorch/circuit.py]: we have to keep it for now i am afraid. resolving
